Issue #218:  Fix decode PATH_INFO on CGI and WSGI

futatuki opened the pull request:
On subversion repository, PATH_INFO is always UTF-8, however on CVS repository, arbitrary nul terminated C string can be used for path.  So we should use bytes or Unicode string with 'surrogateescape' as PATH_INFO, at least before which repository is pointed by the path  is determined.

For WSGI, this was implemented except it is using 'replace' error handler, so this PR change to use 'surrogateescape' handler.

For CGI, PATH_INFO wasn't decoded, this add a code to do it.

cmpilato commented:
I don't mind us declaring that "we should always use UTF-8 for environment variables" as an interface promise (actually, I love promises like that!), but if so then let's a) document that promise in the Server.getenv() stub, and b) make sure the other servers (ModPythonServer, e.g.) adhere to the promise.

futatuki commented:
I looked out getenv method of non-abstract subclasses of the Server class.

- CgiServer
fixed on 685e2ff
- WsgiServer
fixed on 792f409
- AspServer
On Windows platform, as [os.supports_bytes_environment](https://docs.python.org/3/library/os.html#os.supports_bytes_environ) is False, environment variables are always Unicode on Windows platform (it seems ASP itself doesn't support Python 3, though)
- ModPythonServer
It uses [apache.request.subprocess_env](http://modpython.org/live/current/doc-html/pythonapi.html#apache.request.subprocess_env), which is a wrapper of APR tables.  As far as I read the source file ([src/tableobject.c:table_subscript()](https://github.com/grisha/mod_python/blob/master/src/include/mod_python.h.in#L136), [src/include/mod_python.h.in](https://github.com/grisha/mod_python/blob/master/src/include/mod_python.h.in#L136)), it uses PyUnicode_FromString() to get values as Python object, thus it uses always UTF-8.
- StandaloneServer
It uses os.environ, however all environment variables used by viewvc are set by ViewVCHTTPRequestHandler with str value and encoding and errors used by encode and decode are always same. so we need not care what is the encoding.  As ViewVCHTTPRequestHandler unquote value of FILE_PATH before set to it, 
685e2ff caused regression in specific file path which contains '%' character. I'll fix it later.

futatuki commented:
>     * ModPythonServer

I tried to set up mod_python with Python 3.6 on CentOS 7, however I couldn't.  I fixed https://github.com/grisha/mod_python/issues/85 locally,  however make test still failed.
It seems mod_python couldn't find a module specified by PythonHandler directive in httpd.conf.

So I give up confirming the behaivor of the ModPythonServer.getenv farther more.

futatuki commented:
>     * StandaloneServer
>       It uses os.environ, however all environment variables used by viewvc are set by ViewVCHTTPRequestHandler with str value and encoding and errors used by encode and decode are always same. so we need not care what is the encoding.  As ViewVCHTTPRequestHandler unquote value of FILE_PATH before set to it,
>       [685e2ff](https://github.com/viewvc/viewvc/commit/685e2ff60ea424dff8ac6d9891c8e05245a6ab36) caused regression in specific file path which contains '%' character. I'll fix it later.

done in 4757ef9.

cmpilato commented:
> * AspServer
>   On Windows platform, as [os.supports_bytes_environment](https://docs.python.org/3/library/os.html#os.supports_bytes_environ) is False, environment variables are always Unicode on Windows platform (it seems ASP itself doesn't support Python 3, though)

If AspServer doesn't support Python3, let's ditch it altogether.  It can always be re-added later.  (We'll just need to remember to release note its removal.)

cmpilato commented on lib/sapi.py:
```diff
@@ -236,8 +237,25 @@ def redirect(self, url):
+    if name in ('PATH_INFO', 'SCRIPT_NAME') and ret:
+      # Also, we should interpret URL encoding as UTF-8 stream, however,
+      # we should hold whole sequence even if it causes error.
+      ret = _unquote(ret, 'utf-8', 'surrogateescape')
```
I'm pretty sure this is at odds both with the CGI RFC (https://tools.ietf.org/html/rfc3875#section-4.1.5 ; https://tools.ietf.org/html/rfc3875#section-4.1.13) and with common practice under the likes of nginx and Apache HTTP Server.  `PATH_INFO` and `SCRIPT_NAME` are _not_ URI-encoded by the gateway, thus we should not be trying to use `unquote()` on them.

futatuki replied:
I see, I confirmed you are right.  I'll fix it and revert 4757ef9.

futatuki replied:
> I see, I confirmed you are right. I'll fix it and revert [4757ef9](https://github.com/viewvc/viewvc/commit/4757ef995998458dd013c5cf305c395a8a3cacea).

done.

futatuki commented:
> > * AspServer
> >   On Windows platform, as [os.supports_bytes_environment](https://docs.python.org/3/library/os.html#os.supports_bytes_environ) is False, environment variables are always Unicode on Windows platform (it seems ASP itself doesn't support Python 3, though)
> 
> If AspServer doesn't support Python3, let's ditch it altogether. It can always be re-added later. (We'll just need to remember to release note its removal.)

To express more precisely for the fairness, I couldn't find information whether classic ASP supports Python 3 or not.  Perhaps it depends on ActivePython 3.5 still support it or not, but I couldn't see both of positive and negative answer.

Anyways, if none of us can try to run ViewVC on such environments, we can't support them.
So I agree to drop it.

futatuki commented:
> Anyways, if none of us can try to run ViewVC on such environments, we can't support them.
> So I agree to drop it.

I think it is better that the change for it is separated from this PR.

cmpilato commented:
SIDEBAR: I, too, am seeing the same issue with mod_python.

futatuki commented:
It turned out that it is need to specify encoding and 'surrogateescape' error handler expilcitly when we url encode/decode the path, to handle those files that names cannot encode with repositorie's encoding without error in CVS repository.

e.g. URL path  `/EUC-JP%C6%FC%CB%DC%B8%EC.txt` is a representation of the EUC-JP encoded path `/EUC-JP日本語.txt` in CVS repository which encoding is specifed to UTF-8.
`/EUC-JP%C6%FC%CB%DC%B8%EC.txt` can be decoded as '`/EUC-JP\udcc6\udcfc\udccb\udcdc\udcb8\udcec.text`' by 'utf-8' codec with 'surrogateescape' error handler. This path can accessible by vclib/ccvs modules,  although this file name is displayed as `/EUC-JP���ܸ�.txt` on the browser.

futatuki commented:
> It turned out that it is need to specify encoding and 'surrogateescape' error handler expilcitly when we url encode/decode the path, to handle those files that names cannot encode with repositorie's encoding without error in CVS repository.

Done. Now I can access /EUC-JP%C6%FC%CB%DC%B8%EC.txt in CVS repo with ViewVC.

futatuki commented:
Now `/EUC-JP%C6%FC%CB%DC%B8%EC.txt` can be displayed as `/EUC-JP日本語.txt` when the repos' encoding is EUC-JP.

I think this is ready to merge.

futatuki commented:
@cmpilato 
Now I moved the issue about mod_python, then I'll merge this PR to go ahead.
If you think there still exist problems, please revert this merge or open an issue.

cmpilato commented:
ASP support was removed as part of issue #235 .
